src/libcollections/vec.rs

```diff
@@ -989,6 +989,46 @@ impl<T> Vec<T> {
             result
         }
     }
+
+    /// Splits the collection into two at the given index.
+    ///
+    /// Returns a newly allocated `Self`. `self` contains elements `[0..at]`,
+    /// and the returned `Self` contains elements `[at ..]`.
+    ///
+    /// Note that the capacity of `self` does not change.
+    ///
+    /// # Examples
+    /// ```rust
+    /// let mut vec = vec![1,2,3];
+    /// let vec2 = vec.split_off(1);
+    /// assert_eq!(vec, vec![1]);
+    /// assert_eq!(vec2, vec![2, 3]);
+    /// ```
+    #[inline]
+    #[unstable = "new API, waiting for dust to settle"]
+    pub fn split_off(&mut self, at: usize) -> Self {
+        if at >= self.len { panic!("split at larger than vec length") }
+
+        let other_len = self.len - at;
+        let mut other = Vec::with_capacity(other_len);
+        self.len = at;
+
+        // Unsafely `set_len` and conditionally copy items to `other`.
+        unsafe {
+            other.set_len(other_len);
+
+            // Copy items only if `T` is a non-zero sized type.
+            if mem::size_of::<T>() != 0 {
```

Comment:
I don't think you need this guard. This should be a no-op for zero-sized-types.

The only time to you *really* need to guard is when raw allocation is involved, or you want to guarantee a pointer offset has a side-effect (e.g. you want to perform an operation until two pointers are equal).

In this case Rust should just see "oh they want to move 0 bytes somewhere, ok I won't do anything". *Ideally* this would be completely removed during monomorphization, but it certainly shouldn't have any side-effects.